File: Data Structures & Algorithms/binary-tree-diameter/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    int ans = Integer.MIN_VALUE;

    public int dfs(TreeNode root){
        if(root == null){
          return 0;
        }
        int lefth = dfs(root.left);
        int righth = dfs(root.right);

        ans = Math.max(ans,(lefth+righth));
        return 1 + Math.max(lefth,righth);
    }

    public int diameterOfBinaryTree(TreeNode root) {
       int k = dfs(root);
       return ans;
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

    TreeNode* build(vector<int>& preorder, vector<int>& inorder,
                    int preStart, int preEnd,
                    int inStart, int inEnd) {

        if (preStart > preEnd || inStart > inEnd) return NULL;

        int rootVal = preorder[preStart];
        TreeNode* root = new TreeNode(rootVal);

        // 🔴 linear search
        int idx = inStart;
        while (inorder[idx] != rootVal) idx++;

        int leftSize = idx - inStart;

        root->left = build(preorder, inorder,
                           preStart + 1,
                           preStart + leftSize,
                           inStart,
                           idx - 1);

        root->right = build(preorder, inorder,
                            preStart + leftSize + 1,
                            preEnd,
                            idx + 1,
                            inEnd);

        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return build(preorder, inorder, 0, preorder.size()-1, 0, inorder.size()-1);
    }
    
};
